src/main/java/frc/robot/subsystems/Output.java

```diff
@@ -0,0 +1,164 @@
+import com.revrobotics.spark.ClosedLoopSlot;
+import com.revrobotics.spark.SparkBase;
+import com.revrobotics.spark.SparkMax;
+import com.revrobotics.spark.SparkLowLevel.MotorType;
+import com.revrobotics.spark.config.SparkMaxConfig;
+import com.revrobotics.spark.config.SparkBaseConfig.IdleMode;
+import edu.wpi.first.math.geometry.Translation2d;
+import edu.wpi.first.math.system.plant.DCMotor;
+import edu.wpi.first.wpilibj.DigitalInput;
+import frc.robot.Constants;
+import frc.robot.Robot;
+import swervelib.SwerveDrive;
+import static edu.wpi.first.units.Units.Inches;
+import static edu.wpi.first.units.Units.InchesPerSecond;
+import static edu.wpi.first.units.Units.Degrees;
+import com.revrobotics.spark.SparkClosedLoopController;
+import edu.wpi.first.wpilibj2.command.Command;
+import edu.wpi.first.wpilibj2.command.FunctionalCommand;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+
+
+
+public class Output extends SubsystemBase implements AutoCloseable {
+
+  private OutputSimulation m_OutputSim;
+  private LocationService m_LocationService;
+  // declares the motor on the output device
+  private final SparkMax m_OutputMotor =
+      new SparkMax(Constants.OutputConstants.kOutputMotorPort, MotorType.kBrushless);
+
+  // declares the moter gearbox
+  private DCMotor m_OutputGearbox = DCMotor.getNEO(1);
+
+  // declares IR Sensor
+  private final DigitalInput m_coraldetect =
+      new DigitalInput(Constants.OutputConstants.kIRsensorport);
+
+
+  // declares the controller for the output motor
+  // private final SparkClosedLoopController m_Outputcontroller =
+  // m_OutputMotor.getClosedLoopController();
+  // private final RelativeEncoder m_OutputEncoder = m_OutputMotor.getAlternateEncoder();
+  public Output(SwerveDrive drivetrain) {
+
+    // Configure the output motor
+    SparkMaxConfig OutputMotorConfig = new SparkMaxConfig();
+    OutputMotorConfig.smartCurrentLimit(50).idleMode(IdleMode.kBrake);
+
+    OutputMotorConfig.closedLoop.pidf(Constants.OutputConstants.kOutputKp,
+        Constants.OutputConstants.kOutputKi, Constants.OutputConstants.kOutputKd,
+        1.0 / Constants.OutputConstants.kOutputKv, ClosedLoopSlot.kSlot0);
+
+    // Initialize the blank final fields
+    m_OutputSim = new OutputSimulation(m_OutputGearbox);
+
+  }
+
+  // Creates Output Simulation
+
+
+  public void OutputSimulation(DCMotor OutputGearbox) {
+    m_OutputGearbox = OutputGearbox;
+  }
+
+  public class OutputSimulation {
+    private boolean havePiece = false;
+
+    public OutputSimulation(DCMotor OutputGearbox) {
+      m_OutputGearbox = OutputGearbox;
+    }
+
+    public boolean getGamePieceInBot() {
+      return havePiece;
+    }
+
+    public void periodic() {
+      if (m_LocationService.inSourceRegion()) {
+        havePiece = true;
+      }
+    }
+
+    // function to addprojectile, called by ejectCoralCommand in sim mode
+    public void ejectCoralSim() {
+      m_OutputSim.addGamePieceProjectile(null, 0);
```

Comment:
Need this not to be null. driveSimulation gets dereferenced. The constructor for the Output takes a `SwerveDrive`, have it store that somewhere so it can get that and pass it in.

Reply:
Also, the height should be the height of the elevator (might need to get that from the Elevator sim)